In [12]:
%run currency_scrape.ipynb

<Response [200]>


In [13]:
import requests
from bs4 import BeautifulSoup

In [14]:
import pandas as pd
import json
import numpy as np

In [15]:
gameName='Hitman 3'
gameNameAlt='Hitman 3'

---- allkeyshop search----- 

In [16]:
url_endpoint = "https://www.allkeyshop.com/blog/wp-admin/admin-ajax.php"
reqParams = {
    'action': 'catalogue_query',
    'query[search]': gameName,
    'query[productsPerPage]': 16,
    'query[currency]': 'eur',
    'query[category]': 'pc-games-all'
}

resp = requests.get(url_endpoint, params=reqParams).json()['results']

In [17]:
soup = BeautifulSoup(resp, 'html.parser')
productId = soup.find(class_="metacritic-button")['data-product-id']
productId

'38888'

In [18]:
reqParams = {
    'action': 'get_offers',
    'product': productId,
    'currency': 'eur',
    'use_beta_offers_display': 1,
    'region':2
}

payload = {}

response = requests.request(
    "GET", url_endpoint, params=reqParams, data=payload)
#response.json()

In [19]:
file=response.json()

In [20]:
file.keys()

dict_keys(['success', 'offers', 'merchants', 'editions', 'regions'])

In [21]:
file['offers'][0]['price']['eur'].keys()

dict_keys(['bestCoupon', 'currency', 'price', 'priceWithoutCoupon'])

In [22]:
file['offers'][0]['price']['eur']['priceWithoutCoupon']

19.98

In [23]:
data=pd.json_normalize(file,record_path='offers')
data

id                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
2   134524098  https://www.kinguin.net/category/115956/hitman...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
8   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
9   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   
13  134689466  https://store.steampowered.com/app/1659040/HIT...      True   
14  134556470  https://www.kinguin.net/category/118902/hitman...      True   
15  134576917  http://www.gamingdragons.com/en/game/buy-hitma...      True   
16  134638685  https://www.hrkgame.com/en/games/product/hitma...      True   
17  134668487  https://k4g.com/product/hitman-3-steam-global-...      True   
18  134514503  https://www.eneba.com/steam-hitman-3-deluxe-ed...      True   
19  134523817  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
20  134523816  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
21  134658897  https://www.voidu.com/en/product/hitman-3-stan...      True   
22  134658895  https://www.voidu.com/en/product/hitman-3-delu...      True   
23  134567518  https://www.gamers-outlet.net/buy-hitman-3-cd-...      True   
24  134634311  https://www.cdkeys.com/hitman-3-deluxe-edition...      True   

   merchant edition region       stock platform price.eur.bestCoupon.code  \
0       272       1      2     InStock    steam                  8%: AKS8   
1       218       1      2     InStock    steam          8%: AKSpooktober   
2        47       1      2     InStock    steam            14%: OCTOBER14   
3        13       1      2     InStock    steam                       NaN   
4         9       1      2     InStock    steam                       NaN   
5        94       1      2     InStock    steam        -10%: ALLKEYSHOP10   
6        67       1      2     InStock    steam                       NaN   
7       321       1      2     InStock    steam                       NaN   
8       324       1      2     InStock    steam         14%: AKS14PUMPKIN   
9     61616       1      2     InStock    steam         1.99€ Monthly Sub   
10       61       1      2     InStock    steam                       NaN   
11       49       1      2     InStock    steam                       NaN   
12      218       7      2     InStock    steam          8%: AKSpooktober   
13        1       1      2     InStock    steam                       NaN   
14       47       7      2     InStock    steam            14%: OCTOBER14   
15       73       1      2     InStock    steam                       NaN   
16       94       7      2     InStock    steam        -10%: ALLKEYSHOP10   
17      324       7      2     InStock    steam         14%: AKS14PUMPKIN   
18      272       7      2     InStock    steam                  8%: AKS8   
19    61616       7      2     InStock    steam         1.99€ Monthly Sub   
20       61       7      2     InStock    steam                       NaN   
21      222       1      2     InStock    steam                       NaN   
22      222       7      2     InStock    steam                       NaN   
23       77       1      2  OutOfStock    steam                       NaN   
24        9       7      2  OutOfStock    steam          

mapp=data.drop(['price.eur.bestCoupon.discountValue','price.eur.bestCoupon.discountStrategy','price.eur.bestCoupon.isCashback','price.eur.bestCoupon.code', 'price.eur.bestCoupon'], axis=1,inplace=True)

In [24]:
data.rename(columns = {'price.eur.currency':'currency', 'price.eur.price':'price','price.eur.priceWithoutCoupon':'priceWithoutCoupon'}, inplace = True)

In [25]:
data[data.priceWithoutCoupon==data.priceWithoutCoupon.min()]

id                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  priceWithoutCoupon  \
0                           False      eur  18.38               19.98   

   price.eur.bestCoupon  
0                   NaN

---------- dataframe for merchants ------------

In [26]:
merchants_list=pd.DataFrame(data=file['merchants']).T
merchants_list

id            name                 aggregateRating           types  \
272      272           Eneba   {'value': 3.62, 'count': 314}    CD Key Store   
218      218          GAMIVO  {'value': 4.18, 'count': 1248}    CD Key Store   
47        47         Kinguin   {'value': 4.17, 'count': 588}    CD Key Store   
13        13  Instant Gaming   {'value': 4.25, 'count': 307}    CD Key Store   
9          9      CDKeys.com   {'value': 4.17, 'count': 724}    CD Key Store   
94        94             HRK   {'value': 2.71, 'count': 321}    CD Key Store   
67        67      CJS-CDKeys   {'value': 3.81, 'count': 818}    CD Key Store   
321      321   Royal CD Keys    {'value': 4.4, 'count': 289}    CD Key Store   
324      324             K4G   {'value': 4.84, 'count': 675}    CD Key Store   
61616  61616        G2A Plus   {'value': 4.29, 'count': 292}    CD Key Store   
61        61             G2A   {'value': 4.29, 'count': 292}    CD Key Store   
49        49       Play-Asia    {'value': 4.55, 'count': 10}    CD Key Store   
1          1           Steam     {'value': 4.8, 'count': 46}  Official Store   
73        73   Gamingdragons    {'value': 3.04, 'count': 69}    CD Key Store   
222      222        Voidu EU   {'value': 3.89, 'count': 100}  Official Store   
77        77   Gamers Outlet   {'value': 4.09, 'count': 151}    CD Key Store   

                                         paymentMethods        logoSlug  \
272              {'paymentMethode': ['card', 'paypal']}           eneba   
218    {'paymentMethode': ['card', 'paypal', 'crypto']}          gamivo   
47               {'paymentMethode': ['card', 'paypal']}         kinguin   
13               {'paymentMethode': ['paypal', 'card']}  instant-gaming   
9                {'paymentMethode': ['card', 'paypal']}      cdkeys-com   
94     {'paymentMethode': ['card', 'paypal', 'crypto']}             hrk   
67               {'paymentMethode': ['card', 'paypal']}      cjs-cdkeys   
321                        {'paymentMethode': ['card']}   royal-cd-keys   
324              {'paymentMethode': ['card', 'paypal']}             k4g   
61616            {'paymentMethode': ['card', 'paypal']}             g2a   
61               {'paymentMethode': ['card', 'paypal']}             g2a   
49               {'paymentMethode': ['card', 'paypal']}       play-asia   
1                {'paymentMethode': ['card', 'paypal']}           steam   
73               {'paymentMethode': ['card', 'paypal']}   gamingdragons   
222              {'paymentMethode': ['card', 'paypal']}           voidu   
77               {'paymentMethode': ['card', 'paypal']}   gamers-outlet   

                                               reviewUrl  
272        https://www.allkeyshop.com/blog/review/eneba/  
218       https://www.allkeyshop.com/blog/review/gamivo/  
47       https://www.allkeyshop.com/blog/review/kinguin/  
13     https://www.allkeyshop.com/blog/review/instant...  
9      https://www.allkeyshop.com/blog/review/cdkeys_...  
94       https://www.allkeyshop.com/blog/review/hrkgame/  
67     https://www.allkeyshop.com/blog/review/cjs-cdk...  
321    https://www.allkeyshop.com/blog/review/royalcd...  
324          https://www.allkeyshop.com/blog/review/k4g/  
61616        https://www.allkeyshop.com/blog/review/g2a/  
61           https://www.allkeyshop.com/blog/review/g2a/  
49     https://www.allkeyshop.com/blog/review/play-asia/  
1          https://www.allkeyshop.com/blog/review/steam/  
73     https://www.allkeyshop.com/blog/review/gamingd...  
222        https://www.allkeyshop.com/blog/review/voidu/  
77     https://www.allkeyshop.com/blog/review/gamers-...

---- merge merchant names with games hits to make it a bit clearer -----

In [27]:
merged_list = pd.merge(data,merchants_list[['id','name']],left_on='merchant',right_on='id',how='left')
merged_list.drop(['id_y','price'], axis=1,inplace=True)

merged_list.rename(columns = {'priceWithoutCoupon':'price'}, inplace = True)
#merged_list.set_index('id_x',inplace=True)

In [28]:
merged_list.head(-12)

id_x                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
2   134524098  https://www.kinguin.net/category/115956/hitman...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
8   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
9   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   

   merchant edition region    stock platform price.eur.bestCoupon.code  \
0       272       1      2  InStock    steam                  8%: AKS8   
1       218       1      2  InStock    steam          8%: AKSpooktober   
2        47       1      2  InStock    steam            14%: OCTOBER14   
3        13       1      2  InStock    steam                       NaN   
4         9       1      2  InStock    steam                       NaN   
5        94       1      2  InStock    steam        -10%: ALLKEYSHOP10   
6        67       1      2  InStock    steam                       NaN   
7       321       1      2  InStock    steam                       NaN   
8       324       1      2  InStock    steam         14%: AKS14PUMPKIN   
9     61616       1      2  InStock    steam         1.99€ Monthly Sub   
10       61       1      2  InStock    steam                       NaN   
11       49       1      2  InStock    steam                       NaN   
12      218       7      2  InStock    steam          8%: AKSpooktober   

   price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                                8.00                                     %   
1                                8.00                                     %   
2                               14.00                                     %   
3                                 NaN                                   NaN   
4                                 NaN                                   NaN   
5                               10.00                                     %   
6                                 NaN                                   NaN   
7                                 NaN                                   NaN   
8                               14.00                                     %   
9                                0.00                                     %   
10                                NaN                                   NaN   
11                                NaN                                   NaN   
12                               8.00                                     %   

   price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                            False      eur  19.98                   NaN   
1                            False      eur  20.00                   NaN   
2                            False      eur  27.46                   NaN   
3                              NaN      eur  25.75                   NaN   
4                              NaN      eur  25.79                   NaN   
5                            False      eur  29.51                   NaN   
6                              NaN      eur  26.90                   NaN   
7                              NaN      eur  26.94                   NaN   
8                            False      eur  31.66                   NaN   
9           

In [29]:
instock_list=merged_list.loc[(merged_list['platform'] == 'steam') & (merged_list['stock'] == 'InStock')]

In [30]:
instock_list['price'].describe()

count    23.000000
mean     39.804783
std      16.615611
min      19.980000
25%      27.200000
50%      33.500000
75%      51.100000
max      89.990000
Name: price, dtype: float64

In [31]:
table=instock_list[["name", "price"]].groupby("name").min()
table.sort_values('price', ascending=True,inplace=True)

In [32]:
table.head(5)

price
name                 
Eneba           19.98
GAMIVO          20.00
Instant Gaming  25.75
CDKeys.com      25.79
CJS-CDKeys      26.90

In [33]:
instock_list['price'].min()

19.98

In [34]:
instock_list[instock_list.price==instock_list.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon   name  
0                           False      eur  19.98                   NaN  Eneba

In [35]:
instock_list[instock_list.price==instock_list.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon   name  
0                           False      eur  19.98                   NaN  Eneba

In [36]:
allkeysSE = instock_list.assign(priceSEK=instock_list["price"]*rate_sek)
#df2.round({'priceSEK': 1})
allkeysSE.priceSEK=allkeysSE.priceSEK.round(2)
allkeysSE.sort_values(by='priceSEK')

id_x                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
2   134524098  https://www.kinguin.net/category/115956/hitman...      True   
9   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
8   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
13  134689466  https://store.steampowered.com/app/1659040/HIT...      True   
12  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   
15  134576917  http://www.gamingdragons.com/en/game/buy-hitma...      True   
14  134556470  https://www.kinguin.net/category/118902/hitman...      True   
16  134638685  https://www.hrkgame.com/en/games/product/hitma...      True   
19  134523817  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
18  134514503  https://www.eneba.com/steam-hitman-3-deluxe-ed...      True   
20  134523816  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
17  134668487  https://k4g.com/product/hitman-3-steam-global-...      True   
21  134658897  https://www.voidu.com/en/product/hitman-3-stan...      True   
22  134658895  https://www.voidu.com/en/product/hitman-3-delu...      True   

   merchant edition region    stock platform price.eur.bestCoupon.code  \
0       272       1      2  InStock    steam                  8%: AKS8   
1       218       1      2  InStock    steam          8%: AKSpooktober   
3        13       1      2  InStock    steam                       NaN   
4         9       1      2  InStock    steam                       NaN   
6        67       1      2  InStock    steam                       NaN   
7       321       1      2  InStock    steam                       NaN   
2        47       1      2  InStock    steam            14%: OCTOBER14   
9     61616       1      2  InStock    steam         1.99€ Monthly Sub   
5        94       1      2  InStock    steam        -10%: ALLKEYSHOP10   
10       61       1      2  InStock    steam                       NaN   
8       324       1      2  InStock    steam         14%: AKS14PUMPKIN   
11       49       1      2  InStock    steam                       NaN   
13        1       1      2  InStock    steam                       NaN   
12      218       7      2  InStock    steam          8%: AKSpooktober   
15       73       1      2  InStock    steam                       NaN   
14       47       7      2  InStock    steam            14%: OCTOBER14   
16       94       7      2  InStock    steam        -10%: ALLKEYSHOP10   
19    61616       7      2  InStock    steam         1.99€ Monthly Sub   
18      272       7      2  InStock    steam                  8%: AKS8   
20       61       7      2  InStock    steam                       NaN   
17      324       7      2  InStock    steam         14%: AKS14PUMPKIN   
21      222       1      2  InStock    steam                       NaN   
22      222       7      2  InStock    steam                       NaN   

   price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                                8.00                                     %   
1                                8.00                                     %   
3                                 NaN                                   NaN   
4                                 NaN      

#drop crappy - G2A/plus, Eneba
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '61616'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '272'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '61'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '224'].index, inplace = True)
allkeysSE

In [37]:
allkeysSE.loc[allkeysSE['merchant']=='13',['priceSEK','name']].sort_values(by='priceSEK').head(3)

priceSEK            name
3    281.97  Instant Gaming

In [38]:
wanted=['13','63','157','126','22','9']
selectedAllkeysSE=allkeysSE[allkeysSE.merchant.isin(wanted) == True]
selectedAllkeysSE

id_x                                       affiliateUrl  isActive  \
3  134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4  134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
3       13       1      2  InStock    steam                       NaN   
4        9       1      2  InStock    steam                       NaN   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
3                                NaN                                   NaN   
4                                NaN                                   NaN   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
3                             NaN      eur  25.75                   NaN   
4                             NaN      eur  25.79                   NaN   

             name  priceSEK  
3  Instant Gaming    281.97  
4      CDKeys.com    282.41

In [39]:
allkeysSE[allkeysSE.priceSEK==allkeysSE.priceSEK.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                           False      eur  19.98                   NaN   

    name  priceSEK  
0  Eneba    218.78

In [40]:
allkeysSE.loc[allkeysSE['priceSEK'].idxmin()]

id_x                                                                             134513555
affiliateUrl                             https://www.eneba.com/steam-hitman-3-pc-steam-...
isActive                                                                              True
merchant                                                                               272
edition                                                                                  1
region                                                                                   2
stock                                                                              InStock
platform                                                                             steam
price.eur.bestCoupon.code                                                         8%: AKS8
price.eur.bestCoupon.discountValue                                                    8.00
price.eur.bestCoupon.discountStrategy                                                    %

In [41]:
allkeysSE.nsmallest(2,'priceSEK').filter(['name','priceSEK'],axis=1)

name  priceSEK
0   Eneba    218.78
1  GAMIVO    219.00

In [42]:
allkeysSE.loc[allkeysSE['platform']=='steam',['priceSEK','name']].sort_values(by='priceSEK').head(3)

priceSEK            name
0    218.78           Eneba
1    219.00          GAMIVO
3    281.97  Instant Gaming

In [43]:
allkeysSE[allkeysSE.price==allkeysSE.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                           False      eur  19.98                   NaN   

    name  priceSEK  
0  Eneba    218.78